modules/conversantAnalyticsAdapter.js

```diff
@@ -0,0 +1,470 @@
+ */
+export function createAdUnit() {
+  return {
+    sizes: [],
+    bids: {}
+  };
+}
+
+/**
+ * Helper to create a basic bid payload object.  By pre-creating the eventCodes we can easily push in our statuses.
+*/
+export function initializeBidDefaults() {
+  return {
+    'eventCodes': []
+  };
+}
+
+/**
+ * Helper function to send data back to server.  Need to make sure we don't trigger a CORS preflight by not adding
+ * extra header params.
+ * @param payload our JSON payload from either AUCTION END, BID WIN, RENDER FAILED
+ */
+export function sendData(payload) {
+  ajax(URL, function () {}, JSON.stringify(payload), {contentType: 'text/plain'});
+}
+
+/**
+ * Helper to determine if value is integer.  Number.isInteger() not supported everywhere
+ * @param value any value
+ * @returns {boolean} true if integer, false if not
+ */
+export function isInt(value) {
+  if (isNaN(value)) {
+    return false;
+  }
+  const x = parseFloat(value);
+  return (x | 0) === x;
+}
+
+// =============================== BOILERPLATE FOR PRE-BID ANALYTICS SETUP  ============================================
+// save the base class function
+conversantAnalytics.originEnableAnalytics = conversantAnalytics.enableAnalytics;
+conversantAnalytics.originDisableAnalytics = conversantAnalytics.disableAnalytics;
+
+// override enableAnalytics so we can get access to the config passed in from the page
+conversantAnalytics.enableAnalytics = function (config) {
+  if (!config || !config.options || !config.options.site_id) {
+    utils.logError(LOG_PREFIX + 'siteId is required.');
+    return;
+  }
+
+  cacheCleanupInterval = setInterval(
+    function() {
+      const currTime = Date.now();
+      cleanCache(adIdLookup, currTime);
+      cleanCache(timeoutCache, currTime);
+    },
+    CACHE_CLEANUP_TIME_IN_MILLIS
+  );
+
+  initOptions = config.options;
+
+  // Use our default sample rate to determine whether to turn on analytics for this instance. If a sample_rate is defined
+  // in options and it is an integer <= 100 or >= 0 then use that as the sample rate.
+  let sampleRate = DEFAULT_SAMPLE_RATE;
+  if (typeof initOptions.sampleRate !== 'undefined' && isInt(initOptions.sampleRate) &&
+    initOptions.sampleRate >= 0 && initOptions.sampleRate <= 100) {
+    sampleRate = initOptions.sampleRate;
+  }
+  utils.logInfo(LOG_PREFIX + 'Sample rate set to ' + sampleRate + '%');
+  // Math.random() pseudo-random number in the range 0 to less than 1 (inclusive of 0, but not 1)
+  DO_SAMPLE = Math.random() * 100 < sampleRate;
+
+  conversantAnalyticsEnabled = true;
+  conversantAnalytics.originEnableAnalytics(config); // call the base class function
+};
+
+/**
+ * Cleanup code for any timers and caches.
+ */
+conversantAnalytics.disableAnalytics = function () {
+  if (!conversantAnalyticsEnabled) {
+    return;
+  }
+
+  // Cleanup our caches and disable our timer
+  clearInterval(cacheCleanupInterval);
+  cleanCache(timeoutCache, Date.now() + MAX_MILLISECONDS_IN_CACHE);
```

Comment:
Actually no need to call cleanCache.  Much faster to reassign them.

```javascript
timeoutCache = {};
adIdLookup = {};
```

Reply:
Agreed, can't re-assign constants though when I first coded it, now it will work.